packages/coreutils/src/modeldb.ts

```diff
@@ -0,0 +1,611 @@
+ * created and placed in the IModelDB interface.
+ */
+export
+type ObservableType = 'Map' | 'JSON' | 'Vector' | 'UndoableVector' | 'String' | 'Value';
+
+/**
+ * Base interface for Observable objects.
+ */
+export
+interface IObservable extends IDisposable {
+  /**
+   * The type of this object.
+   */
+  readonly type: ObservableType;
+}
+
+/**
+ * Interface for an Observable object that represents
+ * an opaque JSON value.
+ */
+export
+interface IObservableValue extends IObservable {
+  /**
+   * The type of this object.
+   */
+  readonly type: 'Value';
+
+  /**
+   * The changed signal.
+   */
+  readonly changed: ISignal<IObservableValue, ObservableValue.IChangedArgs>;
+
+  /**
+   * Get the current value.
+   */
+  get(): JSONValue;
+
+  /**
+   * Set the value.
+   */
+  set(value: JSONValue): void;
+}
+
+/**
+ * An interface for a path based database for
+ * creating and storing values, which is agnostic
+ * to the particular type of store in the backend.
+ */
+export
+interface IModelDB extends IDisposable {
+  /**
+   * The base path for the `IModelDB`. This is prepended
+   * to all the paths that are passed in to the member
+   * functions of the object.
+   */
+  readonly basePath: string;
+
+  /**
+   * Whether the database has been disposed.
+   */
+  readonly isDisposed: boolean;
+
+  /**
+   * Whether the database has been populated
+   * with model values prior to connection.
+   */
+  readonly isPrepopulated: boolean;
+
+  /**
+   * A promise that resolves when the database
+   * has connected to its backend, if any.
+   */
+  readonly connected: Promise<void>;
+
+  /**
+   * Get a value for a path.
+   *
+   * @param path: the path for the object.
+   *
+   * @returns an `IObservable`.
+   */
+  get(path: string): IObservable;
+
+  /**
+   * Whether the `IModelDB` has an object at this path.
+   *
+   * @param path: the path for the object.
+   *
+   * @returns a boolean for whether an object is at `path`.
+   */
+  has(path: string): boolean;
+
+  /**
+   * Create a string and insert it in the database.
+   *
+   * @param path: the path for the string.
+   *
+   * @returns the string that was created.
+   */
+  createString(path: string): IObservableString;
+
+  /**
+   * Create a vector and insert it in the database.
+   *
+   * @param path: the path for the vector.
+   *
+   * @returns the vector that was created.
+   *
+   * #### Notes
+   * The vector can only store objects that are simple
+   * JSON Objects and primitives.
+   */
+  createVector<T extends JSONValue>(path: string): IObservableVector<T>;
+
+  /**
+   * Create an undoable vector and insert it in the database.
+   *
+   * @param path: the path for the vector.
+   *
+   * @returns the vector that was created.
+   *
+   * #### Notes
+   * The vector can only store objects that are simple
+   * JSON Objects and primitives.
+   */
+  createUndoableVector<T extends JSONValue>(path: string): IObservableUndoableVector<T>;
```

Comment:
Should we just support undo/redo for all object types since they all use JSON values?